docs/CompilerPerformance.md

```diff
@@ -0,0 +1,1115 @@
+     survey of the script above:
+     ```
+     $ utils/scale-test --swiftc-binary=/.../usr/bin/swiftc test.swift.gyb
+     O(n^0.0) : BasicCalleeAnalysis.computeMethodCallees
+     O(n^0.0) : Clang module importer.NumTotalImportedEntities
+     O(n^0.0) : Constraint solver largest system.LargestNumDisjunctionTerms
+     O(n^0.0) : Constraint solver largest system.LargestNumDisjunctions
+     ...
+     O(n^1.0) : regalloc.NumCopies
+     O(n^1.0) : isel.NumFastIselSuccessIndependent
+     O(n^1.0) : regalloc.NumLoads
+     O(n^1.0) : Constraint solver overall.NumSolutionAttempts
+     O(n^1.9) : Constraint solver overall.TotalNumTypeVariables
+     O(n^2.0) : Constraint solver overall.OverallNumComponentsSplit
+     O(n^2.0) : Constraint solver largest system.LargestNumComponentsSplit
+     ...
+     ```
+
+  3. Focus in on a single problematic counter if you see any, and make a
+     regression scale-test -- with a `// RUN: %scale-test` or `// RUN: not
+     %scale-test` line in it depending on whether it's a positive or negative
+     test -- that runs in as little time as possible while still capturing the
+     scaling pattern. Adjust the scaling ranges with `--begin`, `--end`,
+     `--step` and the counters selected with `--select`. It can be helpful to
+     pass `--values` to see the values being extracted. Scale-test uses a
+     numerical nonlinear optimizer that needs at least 4 or 5 data points to fit
+     a curve properly, and will generally do better with 6 or 7 points. By
+     default, `utils/scale-test` will exit `1` (in error) if it fits a scaling
+     curve worse than `O(n^1.2)`, but this threshold can be adjusted up or down
+     to fit the scenario using `--polynomial-threshold` or
+     `--exponential-threshold`.
+
+  4. Add the test to the regression suite. If you're testing a case that you
+     have a fix for, commit the case along with the fix as a positive case, that
+     passes (i.e. shows the bug is fixed). If you're testing a case you do _not_
+     yet have a fix for, commit the case with `// RUN: not %scale-test` and
+     leave it as essentially an XFAIL for others to look at.
+
+  5. If you did _not_ find a problematic counter, it's still important to write
+     and commit a testcase, since it will be able to catch _future_ regressions
+     if the compiler ever starts scaling poorly in the area you've tested. In
+     general, the more scale tests, the more likely performance regressions will
+     be prevented.
+
+  6. If you do not see a counter covering a function in the compiler you're
+     interested in the scaling behaviour of, by all means add a new one!
+     Statistics are easy to add to a file, it takes only a few simple steps:
+     - Add a define like `#define DEBUG_TYPE "subsystem-name"` to the file
+     - Add an include like `#include <swift/Basic/Statistic.h>` to the file
+     - Add `SWIFT_FUNC_STAT;` to the entry of any function you want to measure
```

Comment:
By "entry" you mean as the first (or close-to) statement of the function body, right?

Reply:
Yes, reworded.